File: apps/mobile/lib/features/package/domain/package.dart
enum PackageStatus { received, notified, collected }

PackageStatus _statusFromApi(Object? value) {
  switch (value) {
    case 'NOTIFIED':
      return PackageStatus.notified;
    case 'COLLECTED':
      return PackageStatus.collected;
    default:
      return PackageStatus.received;
  }
}

class Package {
  const Package({
    required this.id,
    required this.recipientName,
    required this.courier,
    required this.trackingNumber,
    required this.status,
    required this.houseCode,
    required this.householdDisplayName,
    required this.receivedAt,
    required this.collectedAt,
    required this.collectedByName,
  });

  final String id;
  final String recipientName;
  final String courier;
  final String? trackingNumber;
  final PackageStatus status;
  final String houseCode;
  final String householdDisplayName;
  final DateTime receivedAt;
  final DateTime? collectedAt;
  final String? collectedByName;

  factory Package.fromJson(Map<String, dynamic> json) {
    final collectedAt = json['collectedAt'];
    return Package(
      id: json['id'] as String,
      recipientName: json['recipientName'] as String,
      courier: json['courier'] as String,
      trackingNumber: json['trackingNumber'] as String?,
      status: _statusFromApi(json['status']),
      houseCode: json['houseCode'] as String,
      householdDisplayName: json['householdDisplayName'] as String,
      receivedAt: DateTime.parse(json['receivedAt'] as String),
      collectedAt: collectedAt is String ? DateTime.parse(collectedAt) : null,
      collectedByName: json['collectedByName'] as String?,
    );
  }
}

File: packages/contracts/src/package.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const packageStatusSchema = z.enum(["RECEIVED", "NOTIFIED", "COLLECTED"]);

export type PackageStatus = z.infer<typeof packageStatusSchema>;

export const createPackageInputSchema = z
  .object({
    houseCode: z.string().trim().min(1).max(24),
    recipientName: z.string().trim().min(2).max(160),
    courier: z.string().trim().min(1).max(80),
    trackingNumber: z.string().trim().max(120).optional(),
  })
  .strict();

export type CreatePackageInput = z.infer<typeof createPackageInputSchema>;

export const collectPackageInputSchema = z
  .object({
    collectedByName: z.string().trim().min(2).max(160),
  })
  .strict();

export type CollectPackageInput = z.infer<typeof collectPackageInputSchema>;

export const packageSchema = z.object({
  id: z.string().uuid(),
  recipientName: z.string().min(1),
  courier: z.string().min(1),
  trackingNumber: z.string().nullable(),
  status: packageStatusSchema,
  houseCode: z.string().min(1),
  householdDisplayName: z.string().min(1),
  receivedAt: z.string().datetime({ offset: true }),
  collectedAt: z.string().datetime({ offset: true }).nullable(),
  collectedByName: z.string().nullable(),
});

export type Package = z.infer<typeof packageSchema>;

export const packageListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(packageSchema) }),
);

export const packageMutationResponseSchema = dataEnvelopeSchema(
  z.object({ package: packageSchema }),
);

export type PackageListResponse = z.infer<typeof packageListResponseSchema>;
export type PackageMutationResponse = z.infer<typeof packageMutationResponseSchema>;
